drivers/nrf24l01 Make the driver work on newer ports

deshipu opened the pull request:
The driver used `pyb` module, which doesn't exist on newer ports,
such as the ESP8266 one. This patch makes two changes:
- Use _FOO with the const() variables, to save some memory.
- Use the `time` module in place of the `pyb` module for delays.
- Use `spi.read`/`spi.write` instead of `spi.send`/`spi.receive`.
- Drop some non-portable parameters to spi and pin initialization.

deshipu committed:
drivers/nrf24l01 Make the driver work on newer ports

The driver used ``pyb`` module, which doesn't exist on newer ports,
such as the ESP8266 one. This patch makes two changes:

* Use the ``utime`` module in place of the ``pyb`` module for delays.
* Use ``spi.read``/``spi.write`` instead of ``spi.send``/``spi.receive``.
* Drop some non-portable parameters to spi and pin initialization.

deshipu commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -48,12 +49,14 @@
         # init the SPI bus and pins
-        spi.init(spi.MASTER, baudrate=4000000, polarity=0, phase=0, firstbit=spi.MSB)
+        spi.init(spi.MASTER, baudrate=4000000, polarity=0, phase=0,
+                 firstbit=spi.MSB)
```
This line is still problematic. ESP8266 doesn't have ``spi.MASTER``, and doesn't accept such an argument to ``spi.init()``.

deshipu commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -48,12 +49,14 @@
+        spi.init(spi.MASTER, baudrate=4000000, polarity=0, phase=0,
+                 firstbit=spi.MSB)
         cs.init(cs.OUT_PP, cs.PULL_NONE)
         ce.init(ce.OUT_PP, ce.PULL_NONE)
```
And those too.

pfalcon commented:
> Use _FOO with the const() variables, to save some memory.

This is controversial change, as now people won't be able to do comfortably things not supported by this driver (by programming registers themselves).

Overall, there's usual problem of "too many changes at once". I'd start with just converting from pyb to machine module (as much as possible) - that's definitely a good move.

deshipu commented:
> Solution to such problems is well-known - dependency injection, i.e. device-specific things should be configured outside reusable code.

I am not sure this well known solution works so well in this case, as the values with which you need to initialize the SPI are also device-specific, but specific to the *other* device, that is, to the one for which the driver is written, and so belong to the driver. Moreover, the driver should re-initialize the SPI to those values every single time it is being used, because in the mean time you might have used a different device on the same SPI bus, which might have set different initialization values. So this definitely belongs in the driver or in some kind of additional layer between the driver and the port's libraries.

deshipu commented:
> This is controversial change, as now people won't be able to do comfortably things not supported by this driver (by programming registers themselves).

I think that people who are comfortable with programming the registers themselves are using the datasheet as a reference anyways, and it's much easier for them to define their own constants based on what they read there, than guess the random names of constants used in the driver, especially since those names are very un-pythonic and use a lot of random abbreviations.

peterhinch commented:
I've used this driver in various applications without needing to program the registers directly. It's quite comprehensive; in my opinion anyone doing so would be skiing well off-piste. The question is whether the saving of 132 bytes warrants giving them a little more leg work. On the ESP I'd tend towards saving the bytes.

deshipu commented:
Any further work needed on this?

dpgeorge commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -1,61 +1,63 @@
-        spi.init(spi.MASTER, baudrate=4000000, polarity=0, phase=0, firstbit=spi.MSB)
-        cs.init(cs.OUT_PP, cs.PULL_NONE)
-        ce.init(ce.OUT_PP, ce.PULL_NONE)
+        spi.init(baudrate=4000000, polarity=0, phase=0)
```
Won't work anymore on pyboard.

dpgeorge commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -1,61 +1,63 @@
-        ce.init(ce.OUT_PP, ce.PULL_NONE)
+        spi.init(baudrate=4000000, polarity=0, phase=0)
+        cs.init(cs.OUT, 1)
+        ce.init(ce.OUT, 0)
```
Should be `value=0`.

dpgeorge commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -99,53 +102,55 @@ def __init__(self, spi, cs, ce, channel=46, payload_size=16):
-        self.spi.send_recv(reg)
-        buf = self.spi.recv(1)
+        self.spi.write(bytearray([reg]))
+        buf = self.spi.read(1)
```
This would definitely benefit from an `spi.write_word` function.

dpgeorge commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -99,53 +102,55 @@ def __init__(self, spi, cs, ce, channel=46, payload_size=16):
-        self.spi.send(buf)
+        response = bytearray(1)
+        self.spi.write_readinto(bytearray([0x20 | reg]), response)
+        status = response[0]
```
This would benefit from `spi.write_word`.

dpgeorge commented:
The conversion of this driver has the same issues (and same solutions) as the SD card driver, #2354.  I think we need an SPI function to transmit/receive a single byte.

dpgeorge commented:
@deshipu to move forward with this PR, can you please:
- revert the change which adds underscore to the start of constants
- make a temporary bytearray in the constructor (likely of 1 byte length) to use for read/write of single bytes
Thanks!

deshipu commented:
I'm sorry for the long delay -- first I didn't see this message, and then I didn't have an opportunity to setup an NRF24L01 and test this properly. Now I have done that, and I incorporated the changes suggested:

* removed the underscores from constants
* used a 1-byte buffer and `spi.readinto`

One thing I'm still not entirely sure of is the use of `utime.ticks_diff` -- there have been some changes and the docs at http://micropython.org/resources/docs/en/latest/esp8266/library/utime.html#utime.ticks_diff are contradictory -- the example doesn't match the text.

dpgeorge commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -198,9 +214,10 @@ def recv(self):
         result = None
-        while result is None and pyb.elapsed_millis(start) < timeout:
+        while (result is None and
+               utime.ticks_diff(start, utime.ticks_ms()) < timeout):
```
The args will need to be swapped, start comes second.

deshipu replied:
Roger.

deshipu commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -98,28 +103,39 @@ def __init__(self, spi, cs, ce, channel=46, payload_size=16):
+        return self.buffer[0]
 
-    def reg_write(self, reg, buf):
+    def reg_write(self, reg, value=None, buffer=None):
```
That's a bad practice (duck typing and all that). I could define a separate method maybe? Something like:

```
def reg_write_bytes(self, reg, bytes):
    ...

def reg_write(self, reg, value):
    self.buffer[0] = value
    return self.reg_write_bytes(reg, self.buffer)
```

dpgeorge replied:
My motivations for the previous comment were 1) not need to make things complex passing keyword args; 2) trying to make this PR as minimal as possible.  If you really think that it's confusing passing either an int or bytearray to the same function then, yes, a separate method would be a better way to handle it.

deshipu commented on drivers/nrf24l01/nrf24l01.py:
```diff
@@ -144,8 +160,8 @@ def set_channel(self, channel):
         assert len(address) == 5
-        self.reg_write(RX_ADDR_P0, address)
-        self.reg_write(TX_ADDR, address)
+        self.reg_write(RX_ADDR_P0, buffer=address)
```
I think it's important to explicitly say when we are sending an int and when we are sending a buffer.

deshipu committed:
Swap ticks_diff arguments

deshipu committed:
Use spearate reg_write and reg_write_bytes functions

Since the buffer is already used to get the status, I couldn't
reuse their code easily, so I decided to duplicate the 4 lines.

deshipu committed:
Add a FEATURE register constant

deshipu committed:
Add ACTIVATE command and FEATURE register

dpgeorge commented:
The repo for this PR seems to have gone missing... so I merged the patch manually and made some changes so the driver doesn't need to allocate on the heap for reading/writing registers.  I also didn't add the additional commits with the feature and activate stuff.  Merged in d7310fabc22ebf45aedcaade1c531ece96c33bf0

pfalcon commented:
> The repo for this PR seems to have gone missing... so I merged the patch manually

Your merge script also relies on the presence of a fork repo? There's another way which I can't find a time to try: http://stackoverflow.com/a/30584951/496009

dpgeorge commented:
> There's another way which I can't find a time to try:

Thanks for the tip!  The following command worked to fetch this PR: `git fetch origin pull/2285/head:nrf24mods`
